Obvius2020

lindavin opened the pull request:
This PR brings in the previously done Obvius2020 pipeline, with the added work of properly creating meters from a .ini configuration file.

Fixes #163.

Note merging this changes the database configuration. Note merging this changes the node modules.

lindavin committed:
Add test for obvius config file processing

lindavin committed:
Wrap config file processing into a function

lindavin committed:
Add returns param

huss committed:
fix obvius bug & upgrade pg-promise

The upgrade of pg-promise is needed per discussion in a recent PR.
This does not completely fix the issues but allows current developers to
work with this. This will be completely resolved once PR is merged into
this branch (after approved).

The upgrade found a mistake I made in DB usage. You should not be doing
a dropConnection() as this does a stopDB() that invalidates DB
connection. pg-promise now detects and errors in this case.
Thus, they were removed.

huss committed:
add better obvius test data

The new test files have several days of data that are hourly.
This means OED can graph the data after refreshing views because
it has multiple days. The switch to hourly data fixes a temporary issue
that the the code assumes each point spans one hour (this needs to be
changed as we merge in the new pipeline code).

It is recommended to look at meter  number 4 as it has the best data.
To see the data reasonably, you need to move the right endpoint and
redraw (we need to look into this issue elsewhere as not related to
obvius). You will lose one day of the three days due to how OED
currently graphs data. Finally, you need to be logged in as admin to see
the meters since display is disabled when obvius meters are added.
You can change this for desired meter(s) on the meter page as admin.

lindavin committed:
Merge remote-tracking branch 'upstream/obvius2020' into obvius2020

lindavin committed:
Replace .param with .body

lindavin committed:
Log warning when an obvius meter is being created from the data file

lindavin committed:
Allow way to create obvius meters from uploaded config file

lindavin committed:
Revert to using the generic parameter pipeline

lindavin committed:
Remove the first three items

We expect that the first three items will just be arrays whose values
are 0.

lindavin committed:
Fix linting errors

lindavin committed:
Replace info with proper warn method

lindavin committed:
Add license to the top of the file

lindavin committed:
Escape html to prevent xss attacks

huss committed:
update packages for what in package.json

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/lindavin/520

huss committed:
basic changes

improve warning message.
update test README for config upload.

huss commented on src/server/models/obvius/processConfigFile.js:
```diff
@@ -0,0 +1,43 @@
+const ConfigFile = require('../../models/obvius/Configfile');
+
+/**
+ * Creates meters from a config file
```
The code actually seems to create an array of meters as the return indicated and not create the meters.

huss commented on src/server/models/obvius/processConfigFile.js:
```diff
@@ -0,0 +1,43 @@
+	// Array of Meter (from models) objects
+	const metersArray = [];
+	for (key of Object.keys(config)) {
+		const [, meterNumber, characteristic] = key.split(regularExpression);
```
I think my knowledge of JS is too limited but I wanted to check that [, is correct. The code does seem to work fine.

huss commented on src/server/routes/obvius.js:
```diff
@@ -0,0 +1,199 @@
+			try {
+				data = zlib.gunzipSync(fx.buffer).toString('utf-8');
+			} catch (error) {
+				data = fx.buffer.toString('utf-8');
```
This is treated differently when the gunzip fails for a log file. Should we be expecting a file that is not zipped? If not, this should be treated as in log upload.

huss commented on src/server/test/web/obviusConfig.js:
```diff
@@ -0,0 +1,78 @@
+const ConfigFile = require('../../models/obvius/Configfile');
+
+const { log } = require('../../log');
+
```
I'm a little unclear about several aspects of the tests in this file. They seem to test functions but then output to the logging system. First, I don't know if doing that is what is desired and is not a good way to test results. Second, I'm not clear what trigger causes the test to fail other than an error. Finally, there are some commented out lines that should either be removed or integrated into the code.

huss commented:
This PR works fine. I want to thank @NoraCodes for important earlier work on the code in this pull request (and @lindavin who has done the recent work). I also note that @mm413 worked on this a little before.

I have merged in the current development. I made a couple of other edits while I did that but they should be minor. There are some issue(s) left that I hope will be resolved when earlier submitted but pending PRs are in development and merged here (waiting to do that). One is that Meter.js mapRow is missing identifier. I will do another round of full testing at that time.

Here are comments on code not changed in the PR that I cannot enter in that file/line:

- src/server/models/obvius/Configfile.js on line 85 has a comment marking the return value uncertain. Is there actually a return value?
- If you upload a config file twice it causes an error of rejected promise. I think an OED error would be best if the meter already exists, etc.
- There are 7 failing tests on my local machine. The first one ("Read Mamac log from a file: 
       rolls back correctly when it rejects:") is likely only on my machine due to a local postgres issue but wanted to know if others see it. The others seem related to obvius work. Can @lindavin  look at these? I can help as needed.

huss reviewed (changes requested):
Just a few files need updating and the general comment I left should be addressed.

huss commented:
I want to note that a few items will need to be addressed but are outside this PR:

- The code assumes every reading is 1 hour. We have had discussions on how to get the correct time for a reading when the start & end times are both not provided. Note this is more complex for the last reading and esp. if there is only one reading provided (as may happen with obvius).
- This code needs to be tried on a live server getting real Obvius data to confirm it works correctly.
- We are working on enhancing the types of users allowed. Once this is done then the password assumed will be removed.

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/lindavin/520

huss committed:
fix commenting mentioned in PR

huss committed:
fix tests

Since the original code was written, we figured out that the first
three columns following the date/time in the CSV file are ignored.
The tests failed because they did not assume this and the current code
does. Test data modified so now correct.

huss committed:
fix meter creation

The new timezone for meters needed to be added as a null argument
to the meter creation.
Note that obviusConfig.js in testing has the same issue but it is not
working in other ways so it should be fixed later (if kept).

huss committed:
remove test that not automated
